sky/packages/sky/lib/src/material/bottom_sheet.dart

```diff
@@ -43,30 +43,41 @@ class _ModalBottomSheetLayout extends OneChildLayoutDelegate {
   }
 }
 
-class _ModalBottomSheetState extends State<_ModalBottomSheet> {
-
-  final _ModalBottomSheetLayout _layout = new _ModalBottomSheetLayout();
+abstract class _BottomSheetDragMixin {
   bool _dragEnabled = false;
 
+  Performance get _performance;
+
   void _handleDragStart(Point position) {
-    _dragEnabled = !config.route._performance.isAnimating;
+    _dragEnabled = !_performance.isAnimating;
   }
 
+  double _dragProgress(double delta);
+
   void _handleDragUpdate(double delta) {
     if (!_dragEnabled)
       return;
-    config.route._performance.progress -= delta / _layout.childTop.end;
+    _performance.progress -= _dragProgress(delta);
   }
 
-  void _handleDragEnd(Offset velocity) {
+  void _handleDragEnd(Offset velocity, BuildContext context) {
     if (!_dragEnabled)
       return;
     if (velocity.dy > _kMinFlingVelocity)
-      config.route._performance.fling(velocity: -velocity.dy * _kFlingVelocityScale);
+      _performance.fling(velocity: -velocity.dy * _kFlingVelocityScale).then((dynamic value) {
+        Navigator.of(context).pop(value);
+      });
     else
-      config.route._performance.forward();
+      _performance.forward();
   }
+}
 
+class _ModalBottomSheetState extends State<_ModalBottomSheet> with _BottomSheetDragMixin {
```

Comment:
Rather than using a mixin, can we use an object that the state holds as a member?  We should eventually share code with Drawer here.  I tried to factor a bit of this logic out of Drawer and into DrawerController.  We can unify with Drawer in a separate step.